### PR TITLE
[Wait for #3094][layer] add Gather Layer

### PR DESCRIPTION
**Pull request by baek2sm**
added gather layer.

- ref: https://pytorch.org/docs/satable/generated/torch.gather.html

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

Signed-off-by: Seungbaek Hong <sb92.hong@samsung.com>

### CODE REVIEW

File: nntrainer/layers/gather_layer.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+void GatherLayer::forwarding_operation(const Tensor &input, const Tensor &index,
+                                       Tensor &output) {
+  for (unsigned int b = 0; b < index.getDim().batch(); ++b) {
+    for (unsigned int i = 0; i < index.getDim().channel(); ++i) {
+      for (unsigned int j = 0; j < index.getDim().height(); ++j) {
+        for (unsigned int k = 0; k < index.getDim().width(); ++k) {
+          auto selected = (size_t)index.getValue(b, i, j, k);
+          if (selected >= input.getDim()[axis]) {
```
Comment (jijoongmoon) [inline nntrainer/layers/gather_layer.cpp:50]: I think it will take so many time. We do need optimize this later. please leave the note. And I wonder what is the difference with concat layer?

Comment (baek2sm) [reply nntrainer/layers/gather_layer.cpp:50]: @jijoongmoon there is no any common ground between concat and gather...

For a 3-D tensor the output is specified by:
out[i][j][k] = input[index[i][j][k]][j][k]  # if dim == 0
out[i][j][k] = input[i][index[i][j][k]][k]  # if dim == 1
out[i][j][k] = input[i][j][index[i][j][k]]  # if dim == 2

gather example 1:
```
>>> t = torch.tensor([[1, 2], [3, 4]])
>>> indices = torch.tensor([[0, 0], [1, 0]])
>>> torch.gather(t, 1, indices)
tensor([[ 1,  1],
        [ 4,  3]])
```

gather example 2:
```
>>> t = torch.tensor([[1, 2, 3], [4, 5, 6], [7, 8, 9]])  
>>> indices = torch.tensor([[0, 2], [0, 1], [0, 2]])
>>> torch.gather(t, 0, indices)
tensor([[1, 8],
        [1, 5],
        [1, 8]])
```
ref: https://pytorch.org/docs/stable/generated/torch.gather.html
